languages/python/sqlalchemy-oso/sqlalchemy_oso/roles2.py

```diff
@@ -0,0 +1,477 @@
+            -- imply a role that has the permission
+            with recursive relevant_roles (role) as (
+                select
+                    role
+                from permission_roles
+                union
+                select
+                    ri.from_role
+                from role_implications ri
+                join relevant_roles rr
+                on ri.to_role = rr.role
+            ) select * from relevant_roles
+        ), user_in_role as (
+            -- check if the user has any of those roles on any of the relevant resources
+            select
+                ur.resource_type,
+                ur.resource_id,
+                ur.role
+            from user_roles ur
+            join relevant_roles rr
+            on rr.role = ur.role
+            join resources r
+            on r.type = ur.resource_type and r.id = ur.resource_id
+            where ur.user_id = :user_id
+        ) select * from user_in_role;
+        """
+
+        self.configured = True
+
+    def _role_allows(self, user, action, resource):
+        # @TODO: Figure out where this session should really come from.
+        assert self.session
+
+        user_pk_name = inspect(user.__class__).primary_key[0].name
+        user_id = getattr(user, user_pk_name)
+
+        resource_pk_name = inspect(resource.__class__).primary_key[0].name
+        resource_id = getattr(resource, resource_pk_name)
+
+        results = self.session.execute(
+            self.sql_query,
+            {
+                "user_id": user.id,
+                "action": action,
+                "resource_id": resource_id,
+                "resource_type": resource.__class__.__name__,
+            },
+        )
+        role = results.first()
+        if role:
+            return True
+        else:
+            return False
+
+    def enable(self):
+        class Roles:
+            @staticmethod
+            def role_allows(user, action, resource):
+                if not self.configured:
+                    self._configure()
+                return self._role_allows(user, action, resource)
+
+        self.oso.register_class(Roles)
+
+    @ensure_configured
+    def assign_role(self, session, user, resource, role_name):
+        # @TODO: Verify all the rules of what roles you can be assigned to.
+        assert role_name in self.roles
+        role = self.roles[role_name]
+
+        assert resource.__class__ in role.python_class
+
+        user_pk_name = inspect(user.__class__).primary_key[0].name
+        user_id = getattr(user, user_pk_name)
+        resource_type = resource.__class__.__name__
+        resource_pk_name = inspect(resource.__class__).primary_key[0].name
+        resource_id = str(getattr(resource, resource_pk_name))
+
+        user_role = self.UserRole(
+            user_id=user_id,
+            resource_type=resource_type,
+            resource_id=resource_id,
+            role=role_name,
+        )
+        session.add(user_role)
```

Comment:
How does conflict resolution work? (User already in role, user has another role on this resource)